### ISSUE TITLE
[WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin

### ISSUE THREAD

**Issue by blc325** [Juniper/ansible-junos-stdlib#209]
I get this when using the actions of ( snap_pre, snap_post, check ) in junos_jsnapy in Ansible. Works fine when using just the standalone jsnapy so I’m thinking it’s ansible related:

 [WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin
(</etc/ansible/roles/Juniper.junos/callback_plugins/jsnapy.CallbackModule object at 0xb68ea42c>): 'final_result'

---
- name: BGP Audit

  roles:
    - Juniper.junos

  hosts: demo

  connection: local
  gather_facts: no

  tasks:

    - name: Do pre maintenance BGP snapshot
      junos_jsnapy:
        host: "{{ inventory_hostname }}"
        user: root
        passwd: juniper1
        test_files: jsnapy/bgp_diff.yml
        action: snap_pre
      register: pre_snapshot

tests_include:
  - check_bgp_status

check_bgp_status:
  - command: show bgp neighbor
  - iterate:
      xpath: //bgp-peer
      tests:
        - no-diff: peer-state

**Comment by blc325** [comment]
jsnapy version (1.0.1)
ansible version (2.1.0.0)

**Comment by stacywsmith** [comment]
@blc325 Can you also provide the version of ansible-junos-stdlib modules and the version of PyEZ? Thanks.

**Comment by blc325** [comment]
JUNOS-eznc is 2.0.1

BTW, I had problems with the jsnapy module that came with Ansible(wouldn't even run the play) so I pulled the latest jsnapy from juniper's Ansible github and placed it in the library of the directory I ran it off of.



On Mar 1, 2017, at 10:00 AM, Stacy W. Smith <notifications@github.com> wrote:

@blc325 Can you also provide the version of ansible-junos-stdlib modules and the version of PyEZ? Thanks.

**Comment by stacywsmith** [comment]
@blc325 Can you please try upgrading to the latest version of jsnapy (1.1) with:
`sudo pip install -U jsnapy`

You may also wish to re-install the Juniper.junos Ansible  role with:
`sudo ansible-galaxy remove Juniper.junos`
`sudo ansible-galaxy install Juniper.junos`

I'm concerned directly copying the junos_jsnapy role file from GitHub may have had a conflict if you didn't also update the jsnapy callback plugin.

**Comment by blc325** [comment]
I did but it did not help...

brianchow@ubuntu:~/ansible/my_demo$ sudo ansible-galaxy list

- Juniper.junos, 1.4.0

brianchow@ubuntu:~/ansible/my_demo$ sudo pip list | grep ansible

*ansible* (2.1.2.0)

brianchow@ubuntu:~/ansible/my_demo$ sudo pip list | grep jsnapy

*jsnapy* (1.1.0)

**Comment by stacywsmith** [comment]
@vnitinv @eeishaan Since your more familiar with JSNAPy, can you please take a look at this?

**Comment by dmontagner** [comment]
@blc325 I don't have a testbed to try this now, so in case you are able to test, can you please try this:

On file `/etc/ansible/roles/Juniper.junos/callback_plugins/jsnapy.py`, change the following:

```
        if res['final_result'] == "Failed":
```

with:

```
        if res['result'] == "Failed":
```

I suspect this may solve the problem.

**Comment by blc325** [comment]
That did not work. Not sure if this helps but final_result is a null value.
From output:


TASK [debug]
*******************************************************************

ok: [192.168.254.254] => {

    "msg": {

        "action": "check",

        "changed": false,

        "final_result": null,

        "passPercentage": 0,

        "router": "192.168.254.254",

        "test_results": {

            "show bgp neighbor": [

                {

                    "count": {

                        "fail": 0,

                        "pass": 0

                    },

                    "failed": [],

                    "node_name": "peer-state",

                    "passed": [],

                    "result": null,

                    "testoperation": "no-diff",

                    "xpath": "//bgp-peer"

                }

            ]

        },

        "total_failed": 0,

        "total_passed": 0,

        "total_tests": 0

    }

}

**Comment by eeishaan** [comment]
`final_result` is null only when none of the test cases are able to run. This can be caused by
1. jsnapy not being able to resolve your testfiles
2. Test cases not being able to run due some absent nodes in snapshot

I suspect former to be the case here. Do you use JSNAPY_HOME environment variable?

**Comment by blc325** [comment]
The test files did run and passed. It's a doing a diff of 2 snapshots and the result is that it did not change. changed==false see below...

msg": {

        "action": "check", 

        "changed": false, 

        "final_result": null, 

        "passPercentage": 0, 

        "router": "192.168.254.254", 




On Mar 4, 2017, at 3:16 AM, Ishaan Kumar <notifications@github.com> wrote:

final_result is null only when none of the test cases are able to run. This can be caused by

jsnapy not being able to resolve your testfiles
Test cases not being able to run due some absent nodes in snapshot
I suspect former to be the case here. Do you use JSNAPY_HOME environment variable?

**Comment by eugeneee** [comment]
@blc325 you'll need to replace jsnapy.py with this https://github.com/Juniper/ansible-junos-stdlib/blob/master/callback_plugins/jsnapy.py 

Please replace at the following directory:
/etc/ansible/roles/Juniper.junos/callback_plugins

This will help. I've got it running.

**Comment by vnitinv** [comment]
@blc325  @eugeneee suggestion should get you going. If you still feel the problem, I will look into this issue.

**Comment by blc325** [comment]
Hi Nitin,

I tried replacing the file but I still got the error:


TASK [debug]
*******************************************************************

ok: [192.168.254.254] => {

    "msg": {

        "action": "check",

        "changed": false,

        "final_result": null,

        "passPercentage": 0,

        "router": "192.168.254.254",

        "test_results": {

            "show bgp neighbor": [

                {

                    "count": {

                        "fail": 0,

                        "pass": 0

                    },

                    "failed": [],

                    "node_name": "peer-state",

                    "passed": [],

                    "result": null,

                    "testoperation": "no-diff",

                    "xpath": "//bgp-peer"

                }

            ]

        },

        "total_failed": 0,

        "total_passed": 0,

        "total_tests": 0

    }

}


TASK [Check Results]
***********************************************************

ok: [192.168.254.254]


PLAY RECAP
*********************************************************************

192.168.254.254            : ok=6    changed=2    unreachable=0
failed=0


* [WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin*

*(</etc/ansible/roles/Juniper.junos/callback_plugins/bak.jsnapy.CallbackModule
object at 0xb696e9ac>): 'final_result'*

**Comment by rahkumar651991** [comment]
Closing the issue as fix merged and released in 2.4.0 roles.
